drivers/wifi/esp8266/esp8266_offload.c

```diff
@@ -0,0 +1,737 @@
+					    NULL, 0U, false);
+	k_sem_give(&dev->cmd_handler_data.sem_tx_lock);
+
+	net_pkt_unref(sock->tx_pkt);
+	sock->tx_pkt = NULL;
+
+	return ret;
+}
+
+static void esp8266_send_work(struct k_work *work)
+{
+	struct esp8266_socket *sock;
+	struct net_context *context;
+	net_context_send_cb_t cb;
+	struct esp8266_data *dev;
+	void *user_data;
+	int ret = 0;
+
+	sock = CONTAINER_OF(work, struct esp8266_socket, send_work);
+	dev = esp8266_socket_to_dev(sock);
+
+	esp8266_lock(dev);
+
+	cb = sock->send_cb;
+	context = sock->context;
+	user_data = sock->send_user_data;
+
+	ret = _sock_send(dev, sock);
+	if (ret < 0) {
+		LOG_ERR("Failed to send data: link %d, ret %d", sock->link_id,
+			ret);
+	}
+
+	esp8266_unlock(dev);
+
+	if (cb) {
+		cb(context, ret, user_data);
+	}
+}
+
+static int esp8266_sendto(struct net_pkt *pkt,
+			  const struct sockaddr *dst_addr,
+			  socklen_t addrlen,
+			  net_context_send_cb_t cb,
+			  s32_t timeout,
+			  void *user_data)
+{
+	struct net_context *context;
+	struct esp8266_socket *sock;
+	struct esp8266_data *dev;
+	int ret = 0;
+
+	context = pkt->context;
+	sock = (struct esp8266_socket *)context->offload_context;
+	dev = esp8266_socket_to_dev(sock);
+
+	LOG_DBG("link %d, timeout %d", sock->link_id, timeout);
+
+	esp8266_lock(dev);
+
+	if (sock->tx_pkt) {
+		esp8266_unlock(dev);
+		return -EBUSY;
+	}
+
+	if (sock->type == SOCK_STREAM) {
+		if (!esp8266_socket_connected(sock)) {
+			esp8266_unlock(dev);
+			return -ENOTCONN;
+		} else if (dst_addr) {
+			esp8266_unlock(dev);
+			return -EISCONN;
+		}
+	} else {
+		if (!esp8266_socket_connected(sock)) {
+			if (!dst_addr) {
+				esp8266_unlock(dev);
+				return -ENOTCONN;
+			}
+
+			/* Use a timeout of 5000 ms here even though the
+			 * timeout parameter might be different. We want to
+			 * have a valid link id before proceeding.
+			 */
+			ret = esp8266_connect(context, dst_addr, addrlen, NULL,
+					      K_SECONDS(5), NULL);
+			if (ret < 0) {
+				esp8266_unlock(dev);
+				return ret;
+			}
+		} else if (dst_addr && memcmp(dst_addr, &sock->dst, addrlen)) {
+			/* This might be unexpected behaviour but the ESP8266
+			 * doesn't support changing endpoint.
+			 */
+			esp8266_unlock(dev);
+			return -EISCONN;
+		}
+	}
+
+	sock->tx_pkt = pkt;
+	sock->send_cb = cb;
+	sock->send_user_data = user_data;
+
+	if (timeout == K_NO_WAIT) {
+		k_work_submit_to_queue(&dev->workq, &sock->send_work);
+		esp8266_unlock(dev);
+		return 0;
+	}
+
+	/*
+	 * FIXME:
+	 * In _modem_cmd_send() in modem_cmd_handler.c it can happen that a
+	 * response, ie 'OK', is received before k_sem_reset(sem) is called.
+	 * If the sending thread can be preempted, the command handler could
+	 * run and call k_sem_give(). This will cause a timeout and the send
+	 * will fail. This can be avoided by locking the scheduler. Maybe this
+	 * should be done in _modem_cmd_send() instead.
+	 */
+	k_sched_lock();
+	ret = _sock_send(dev, sock);
+	k_sched_unlock();
```

Comment:
@mike-scott Have you ever seen the problem I'm describing here? I guess the `modem_cmd_send()` usually is called from a co-op thread in which case this is no problem. Do you think it makes sense to have the `k_sched_lock/unlock` in the `_modem_cmd_send()`?